File: LeetCode-Java/Partition List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode leftHead = null,leftTail = null, rightHead = null, rightTail = null;
        while(head != null)
        {
            if(head.val < x)
            {
                if(leftHead == null)
                {
                    leftHead = head;
                    leftTail = head;
                }
                else
                {
                    leftTail.next = head;
                    leftTail = head;
                }
            }
            else
            {
                if(rightHead == null)
                {
                    rightHead = head;
                    rightTail = head;
                }
                else
                {
                    rightTail.next = head;
                    rightTail = head;
                }
            }
            head = head.next;
        }
        if(rightTail != null)
            rightTail.next = null;
        if(leftHead == null) 
            return rightHead;
        else
        {
            leftTail.next = rightHead;
            return leftHead;
        }
    }
}

File: Leetcode-cpp/partition_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* fake_less = new ListNode(-1);
        ListNode* index_less = fake_less;
        ListNode* fake_greater = new ListNode(-1);
        ListNode* index_greater = fake_greater;
        while (head != nullptr) {
            if (head->val < x) {
                index_less->next = head;
                index_less = head;
            } else {
                index_greater->next = head;
                index_greater = head;
            }
            head = head->next;
        }
        index_less->next = fake_greater->next;
        index_greater->next = nullptr;
        return fake_less->next;
    }
};